[SPARK-14400] [SQL] ScriptTransformation does not fail the job for bad user command

tejasapatil opened the pull request:
## What changes were proposed in this pull request?
- Refer to the Jira for the problem: jira : https://issues.apache.org/jira/browse/SPARK-14400
- The fix is to check if the process has exited with a non-zero exit code in `hasNext()`. I have moved this and checking of writer thread exception to a separate method.
## How was this patch tested?
- Ran a job which had incorrect transform script command and saw that the job fails
- Existing unit tests for `ScriptTransformationSuite`. Added a new unit test

tejasapatil committed:
- Moved the failure check to a method
- Added test case

tejasapatil committed:
Removed unwanted changes

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #55074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55074/consoleFull)** for PR 12194 at commit [`df48e1e`](https://github.com/apache/spark/commit/df48e1e27e0ab569a863da2f9f79cddcbee8169b).

SparkQA commented:
**[Test build #55074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55074/consoleFull)** for PR 12194 at commit [`df48e1e`](https://github.com/apache/spark/commit/df48e1e27e0ab569a863da2f9f79cddcbee8169b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55074/
Test PASSed.

rxin commented:
cc @srowen

tejasapatil committed:
Made the test case more robust

SparkQA commented:
**[Test build #55358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55358/consoleFull)** for PR 12194 at commit [`e3899c3`](https://github.com/apache/spark/commit/e3899c3fd216b33668092a84f595c6d481df1531).

SparkQA commented:
**[Test build #55358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55358/consoleFull)** for PR 12194 at commit [`e3899c3`](https://github.com/apache/spark/commit/e3899c3fd216b33668092a84f595c6d481df1531).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55358/
Test FAILed.

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #55360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55360/consoleFull)** for PR 12194 at commit [`ebb5ea1`](https://github.com/apache/spark/commit/ebb5ea1a5f0cc25fa6112b81a8bf84cec01a24bf).

SparkQA commented:
**[Test build #55360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55360/consoleFull)** for PR 12194 at commit [`ebb5ea1`](https://github.com/apache/spark/commit/ebb5ea1a5f0cc25fa6112b81a8bf84cec01a24bf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55360/
Test FAILed.

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #55676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55676/consoleFull)** for PR 12194 at commit [`1054b71`](https://github.com/apache/spark/commit/1054b717859d37b8c0dd7b41a087cd2924b97b0a).

SparkQA commented:
**[Test build #55676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55676/consoleFull)** for PR 12194 at commit [`1054b71`](https://github.com/apache/spark/commit/1054b717859d37b8c0dd7b41a087cd2924b97b0a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55676/
Test FAILed.

tejasapatil committed:
rebased

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #58286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58286/consoleFull)** for PR 12194 at commit [`e37e0aa`](https://github.com/apache/spark/commit/e37e0aaf9291c4cf71c54cff93e0d341bd80595d).

SparkQA commented:
**[Test build #58286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58286/consoleFull)** for PR 12194 at commit [`e37e0aa`](https://github.com/apache/spark/commit/e37e0aaf9291c4cf71c54cff93e0d341bd80595d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58286/
Test FAILed.

tejasapatil committed:
check for `Exception` instead of `SparkException` since the writer thread might be the first to see the effect of process being killed

```
Exception in thread "Thread-ScriptTransformation-Feed" java.io.IOException: Broken pipe
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:326)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread$$anonfun$run$1.apply$mcV$sp(ScriptTransformation.scala:307)
	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread$$anonfun$run$1.apply(ScriptTransformation.scala:268)
	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread$$anonfun$run$1.apply(ScriptTransformation.scala:268)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1793)
	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread.run(ScriptTransformation.scala:268)
```

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #58720 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58720/consoleFull)** for PR 12194 at commit [`524eb71`](https://github.com/apache/spark/commit/524eb71f2546062ea9c480c6046313be39bbe342).

SparkQA commented:
**[Test build #58720 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58720/consoleFull)** for PR 12194 at commit [`524eb71`](https://github.com/apache/spark/commit/524eb71f2546062ea9c480c6046313be39bbe342).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58720/
Test FAILed.

tejasapatil committed:
Handle exceptions in the ScriptTransformationWriterThread

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #58750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58750/consoleFull)** for PR 12194 at commit [`abd65d8`](https://github.com/apache/spark/commit/abd65d8587e921c02267c889da99169d53b17398).

SparkQA commented:
**[Test build #58750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58750/consoleFull)** for PR 12194 at commit [`abd65d8`](https://github.com/apache/spark/commit/abd65d8587e921c02267c889da99169d53b17398).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58750/
Test PASSed.

tejasapatil commented:
Can anyone please review this PR ?

tejasapatil commented:
Can anyone please review this PR ?

sameeragarwal commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala:
```diff
@@ -127,45 +127,78 @@ case class ScriptTransformation(
+            if (exitCode != 0) {
+              logError(stderrBuffer.toString) // log the stderr circular buffer
+              throw new SparkException(s"Subprocess exited with status $exitCode. " +
+                s"Error: ${stderrBuffer.toString}", cause)
```
unrelated to this PR but we should probably only print the partial contents of the circular buffer if the number of bytes written to it are less than its total size

tejasapatil replied:
@sameeragarwal : Thanks for pointing that out. I will submit a separate PR for that.

tejasapatil replied:
here : https://github.com/apache/spark/pull/13351

sameeragarwal commented:
This patch looks good to me. @srowen / @JoshRosen  can one of you please take a second look as well?

rxin commented:
@srowen want to review this? Given you reviewed the last transform pr.

rxin commented:
@tejasapatil on a related note, we'd want to remove as much hive code dependency as possible. One command that is left is ScriptTransformation. Would you have time to implement this in sql/core without Hive dependency for ScriptTransformation? It seems like one of the primary operators you guys would use.

srowen commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala:
```diff
@@ -127,45 +127,78 @@ case class ScriptTransformation(
+                  checkFailureAndPropagate()
+                  false
+                } else {
+                  true
```
Hm, is it me or does this get hard to follow the return values. Generally the method returns "true" unless one of several conditions caused it to decide it was finished earlier. Those could be handled with early "return false" rather than lots of "else ... true" branches.

rxin replied:
yea i think that's a great idea. the level of nesting is a little bit too much here

tejasapatil replied:
+1. Did the change

srowen commented:
With one minor suggestion this LGTM

rxin commented:
@tejasapatil want to update this so we can merge it for 2.0?

tejasapatil committed:
simplified the `if else` nestedness

SparkQA commented:
**[Test build #59466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59466/consoleFull)** for PR 12194 at commit [`4a7b2e8`](https://github.com/apache/spark/commit/4a7b2e8827ade4b999206454092101139a92895d).

tejasapatil commented:
@rxin : Sorry for delay... was caught up in some other things. I have updated the PR now with the review comments. 

Re your suggestion about removing hive code dependency: I will work on it

sameeragarwal commented:
LGTM pending jenkins. Thanks!

SparkQA commented:
**[Test build #59466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59466/consoleFull)** for PR 12194 at commit [`4a7b2e8`](https://github.com/apache/spark/commit/4a7b2e8827ade4b999206454092101139a92895d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59466/
Test PASSed.

rxin commented:
Merging in master/2.0.
